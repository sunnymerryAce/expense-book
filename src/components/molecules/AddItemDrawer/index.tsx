import React, { useState } from "react";
import Box from "@material-ui/core/Box";
import Container from "@material-ui/core/Container";
import SwipeableDrawer from "@material-ui/core/SwipeableDrawer";
import TextInput from "@/components/atoms/TextInput";
import DateInput from "@/components/atoms/DateInput";
import H6Title from "@/components/atoms/H6Title";
import CloseButton from "@/components/atoms/CloseButton";
import CompleteButton from "@/components/atoms/CompleteButton";
import DeleteButton from "@/components/atoms/DeleteButton";
import { Categories } from "@/state/categories";
import CategorySelector from "@/components/molecules/CategorySelector";
import useStyles from "./style";

export interface AddItemDrawerProps {
  categories: Categories;
  title: string;
  isEditItem: boolean;
  isOpen: boolean;
  toggleDrawer: (props: boolean) => void;
  handleChangeCategory: (...props: any[]) => any;
  handleChangeItemName: (...props: any[]) => any;
  handleChangePrice: (...props: any[]) => any;
  handleChangeDate: (...props: any[]) => any;
}

const AddItemDrawer: React.FC<AddItemDrawerProps> = (props) => {
  const classes = useStyles();
  const [itemNameError, setItemNameError] = useState<boolean>(false);
  const [priceError, setPriceError] = useState<boolean>(false);

  return (
    <SwipeableDrawer
      anchor="bottom"
      open={props.isOpen}
      onClose={() => {
        props.toggleDrawer(false);
      }}
      onOpen={() => {
        props.toggleDrawer(true);
      }}
    >
      <Box className={classes.closeButtonWrapper}>
        <CloseButton
          handleClick={() => {
            props.toggleDrawer(false);
          }}
        ></CloseButton>
      </Box>
      <Container maxWidth="sm">
        <H6Title text={props.title}></H6Title>
        <Box className={classes.categorySelectorWrapper}>
          <CategorySelector
            categories={props.categories}
            handleChangeCategory={props.handleChangeCategory}
          ></CategorySelector>
        </Box>
        <Box className={classes.inputArea}>
          <TextInput
            label="アイテム名"
            handleChange={props.handleChangeItemName}
            error={itemNameError}
            helperText={itemNameError ? "入力してください。" : ""}
            className="ItemNameInput"
          ></TextInput>
          <TextInput
            label="金額"
            type="number"
            handleChange={props.handleChangePrice}
            helperText={priceError ? "入力してください。" : ""}
            error={priceError}
            className="PriceInput"
          ></TextInput>
          <DateInput handleChange={props.handleChangeDate}></DateInput>
        </Box>
      </Container>
      <Box className={classes.completeButtonWrapper}>
        <CompleteButton
          handleClick={() => {
            props.toggleDrawer(false);
          }}
        ></CompleteButton>
      </Box>
      {props.isEditItem && (
        <Box className={classes.deleteButtonWrapper}>
          <DeleteButton
            handleClick={() => {
              props.toggleDrawer(false);
            }}
          ></DeleteButton>
        </Box>
      )}
    </SwipeableDrawer>
  );
};

export default AddItemDrawer;
