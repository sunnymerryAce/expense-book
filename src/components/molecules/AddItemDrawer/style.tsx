import { makeStyles, createStyles, Theme } from "@material-ui/core/styles";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      paddingBottom: "20px",
    },
    closeButtonWrapper: {
      position: "absolute",
      top: "8px",
      left: "0px",
    },
    completeButtonWrapper: {
      position: "fixed",
      right: "20px",
      bottom: "20px",
    },
    deleteButtonWrapper: {
      position: "fixed",
      right: "92px",
      bottom: "20px",
    },
    inputFlexContainer: {
      display: "flex",
      justifyContent: "space-between",
      flexWrap: "wrap",
    },
  })
);

export default useStyles;
