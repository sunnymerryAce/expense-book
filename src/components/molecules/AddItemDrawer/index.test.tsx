import React from "react";
import AddItemDrawer from "./";
import { shallow } from "enzyme";

describe("<AddItemDrawer />", () => {
  let container: any = null;
  const categories = {
    Food: {
      name: "Food",
      color: "#7CB342",
      defaultBudget: 30000,
    },
    Cafe: {
      name: "Cafe",
      color: "#D81B60",
      defaultBudget: 5000,
    },
    雑費: {
      name: "雑費",
      color: "#FDD835",
      defaultBudget: 12000,
    },
    Drink: {
      name: "Drink",
      color: "#5E35B1",
      defaultBudget: 45000,
    },
    Date: {
      name: "Date",
      color: "#FB8C00",
      defaultBudget: 20000,
    },
    Book: {
      name: "Book",
      color: "#1E88E5",
      defaultBudget: 3000,
    },
    Gym: {
      name: "Gym",
      color: "#F4511E",
      defaultBudget: 12000,
    },
    Fixed: {
      name: "Fixed",
      color: "#00ACC1",
      defaultBudget: 33000,
    },
    Sudden: {
      name: "Sudden",
      color: "#8E24AA",
      defaultBudget: 30000,
    },
    Savings: {
      name: "Savings",
      color: "#3949AB",
      defaultBudget: 45000,
    },
  };

  const props = {
    categories: categories,
    title: "New Item",
    isEditItem: false,
    isOpen: true,
    toggleDrawer: jest.fn(),
    handleChangeCategory: jest.fn(),
    handleChangeItemName: jest.fn(),
    handleChangePrice: jest.fn(),
    handleChangeDate: jest.fn(),
  };

  afterEach(() => {
    container.unmount();
    container = null;
  });

  test("should match the snapshot", () => {
    container = shallow(<AddItemDrawer {...props} />);
    expect(container.html()).toMatchSnapshot();
  });

  test("should have proper title", () => {
    container = shallow(<AddItemDrawer {...props} />);
    expect(container.find("H6Title").props().text).toBe(props.title);
  });

  test("should not have delete button on add mode", () => {
    container = shallow(<AddItemDrawer {...props} />);
    expect(container.find("DeleteButton").exists()).toBe(false);
  });

  test("should have delete button on edit mode", () => {
    props.isEditItem = true;
    container = shallow(<AddItemDrawer {...props} />);
    expect(container.find("DeleteButton").exists()).toBe(true);
  });

  test("should have proper handler - CategorySelector", () => {
    container = shallow(<AddItemDrawer {...props} />);
    expect(container.find("CategorySelector").props()).toMatchObject({
      categories: props.categories,
      handleChangeCategory: props.handleChangeCategory,
    });
  });

  test("should have proper handler - ItemNameInput", () => {
    container = shallow(<AddItemDrawer {...props} />);
    expect(container.find(".ItemNameInput").props()).toMatchObject({
      handleChange: props.handleChangeItemName,
    });
  });

  test("should have proper handler - PriceInput", () => {
    container = shallow(<AddItemDrawer {...props} />);
    expect(container.find(".PriceInput").props()).toMatchObject({
      handleChange: props.handleChangePrice,
    });
  });

  test("should have proper handler - DateInput", () => {
    container = shallow(<AddItemDrawer {...props} />);
    expect(container.find("DateInput").props()).toMatchObject({
      handleChange: props.handleChangeDate,
    });
  });
});
