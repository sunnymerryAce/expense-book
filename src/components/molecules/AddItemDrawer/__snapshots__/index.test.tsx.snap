// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddItemDrawer /> should match the snapshot 1`] = `""`;
